### PR TITLE
Analytics V0.0.1. Tested with single_stream

### PR DESCRIPTION
**Pull request by Pixcell**

### CODE REVIEW

File: src/darcyai/pipeline.py [hunk 1]
```diff
@@ -757,7 +801,15 @@ def run(self) -> None:
                     self.__logger.exception("Error running Pipeline")
                     if self.__input_stream_error_handler_callback is not None:
                         self.__input_stream_error_handler_callback(e)
+                        try:
+                            reporter.on_pipeline_error(e, False)
+                        except:
+                            pass
```
Comment (baghbidi) [inline src/darcyai/pipeline.py:807]: These try..except blocks are a little confusing... May be move all the reporter stuff to methods that ignore exceptions and call those methods here in the `pipeline.run`

File: src/darcyai/reporter/analytics_reporter.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+class PipelineBaseEvent():
+  def __init__(self,
+              machine_id: str,
+              pipeline_config_hash: str,
+              pipeline_run_uuid: str):
+    self.machine_id = machine_id
+    self.pipeline_config_hash = pipeline_config_hash
+    self.pipeline_run_uuid = pipeline_run_uuid
+    return
```
Comment (baghbidi) [inline src/darcyai/reporter/analytics_reporter.py:32]: Not needed

File: src/darcyai/reporter/analytics_reporter.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+      hosts.extend(hostnames)
+  return hosts
+
+def __is_using_iofog(self):
+  """
+  Verifies if iofog is present as an extra host
+  """
+  hosts = __get_etc_hostnames()
+  for host in hosts:
+    print(host)
```
Comment (baghbidi) [inline src/darcyai/reporter/analytics_reporter.py:139]: Remove

File: src/darcyai/reporter/analytics_reporter.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+  def __cancel_heartbeat(self):
+    try:
+      if self.__heartbeat_tread is not None:
+        self.__heartbeat_running = False
+        self.__heartbeat_thread.join()
+        self.__heartbeat_tread = None
+      return
+    except AttributeError as e:
+      return
```
Comment (baghbidi) [inline src/darcyai/reporter/analytics_reporter.py:192]: `pass`

**Review by baghbidi — CHANGES_REQUESTED** [review]

**Review by baghbidi — APPROVED** [review]

File: src/darcyai/pipeline.py [hunk 5]
```diff
@@ -74,6 +78,7 @@ class Pipeline():
         the REST API. Defaults to `None`.
     rest_api_port (int): The port of the REST API. Defaults to `5000`.
     rest_api_host (str): The host of the REST API. Defaults to `localhost`.
+    disable_reporting (bool): Disables anonymous usage collection. Defaults to `False`.
```
Comment (neilotoole) [inline src/darcyai/pipeline.py:81]: change to `disable_telemetry`

File: src/darcyai/pipeline.py [hunk 6]
```diff
@@ -113,7 +119,8 @@ def __init__(self,
                  rest_api_base_path: str = None,
                  rest_api_flask_app: Flask = None,
                  rest_api_port: int = None,
-                 rest_api_host: str = None):
+                 rest_api_host: str = None,
+                 disable_reporting: bool = False,):
```
Comment (neilotoole) [inline src/darcyai/pipeline.py:123]: change to `disable_telemetry`

File: src/darcyai/pipeline.py [hunk 7]
```diff
@@ -138,6 +145,9 @@ def __init__(self,
             validate(callable(input_stream_error_handler_callback),
                      "input_stream_error_handler_callback must be a function")
 
+        if disable_reporting is not None:
```
Comment (neilotoole) [inline src/darcyai/pipeline.py:148]: change to `disable_telemetry`

File: src/darcyai/pipeline.py [hunk 8]
```diff
@@ -214,6 +224,10 @@ def __init__(self,
 
         self.__running = False
 
+        # For analytics purposes
+        self.__disable_reporting = disable_reporting
```
Comment (neilotoole) [inline src/darcyai/pipeline.py:228]: change to `disable_telemetry`

File: src/darcyai/pipeline.py [hunk 9]
```diff
@@ -745,7 +759,56 @@ def run(self) -> None:
 
         perceptors_order = self.__get_perceptors_order()
 
+        reporter = AnalyticsReporter(darcyai_version, self.__disable_reporting)
```
Comment (neilotoole) [inline src/darcyai/pipeline.py:762]: Change from `AnalyticsReporter` to `TelemetryReporter`.

File: src/darcyai/reporter/analytics_reporter.py [hunk 10]
```diff
@@ -0,0 +1,416 @@
+import analytics
```
Comment (neilotoole) [inline src/darcyai/reporter/analytics_reporter.py:1]: Change package/module/file from `src/darcyai/reporter/analytics_reporter.py` to `src/darcyai/telemetry/telemetry_reporter.py`. Basically, `telemetry` is now the preferred name for `analytics`.

**Review by neilotoole — CHANGES_REQUESTED** [review]
Initial pass, looks good to me.

Main change is to favor the name `telemetry` instead of `analytics` or `reporting`. It's a subtle change in emphasis, but that's the one we're going with. Sounds less creepy.

File: src/darcyai/pipeline.py [hunk 11]
```diff
@@ -95,7 +100,8 @@ class Pipeline():
     ...                     rest_api_base_path="/",
     ...                     rest_api_flask_app=None,
     ...                     rest_api_port=5000,
-    ...                     rest_api_host="localhost")
+    ...                     rest_api_host="localhost",
+    ...                     disable_reporting=False)
```
Comment (neilotoole) [inline src/darcyai/pipeline.py]: change to `disable_telemetry`

**Pull request closed without merging**